ios/KaMPKitiOS/ViewController.swift

```diff
@@ -15,12 +15,20 @@ class BreedsViewController: UIViewController {
     var data: [Breed] = []
     
     let log = koin.get(objCClass: Kermit.self, parameter: "ViewController") as! Kermit
-    
+
     lazy var adapter: NativeViewModel = NativeViewModel(
-        viewUpdate: { [weak self] summary in
-            self?.viewUpdate(for: summary)
-        }, errorUpdate: { [weak self] errorMessage in
-            self?.errorUpdate(for: errorMessage)
+        onLoading: {
+            // Show a loading spinner
+        },
+        onSuccess: { [weak self] summary in
+            self?.viewUpdateSuccess(for: summary)
+        },
+        onError: {
+            [weak self] error in
+            self?.errorUpdate(for: (error as! Error).localizedDescription)
```

Comment:
I personally avoid force unwrapping pretty much whenever I can. I arguably avoid it too much, but I think it's a relatively common habit among Swift devs. If for some unthinkable reason this couldn't be cast as an `Error`, then the app would crash on this line.

I also think we should keep the pattern of each Swift lambda being a single line, and doing all the work in the function we pass in. Just to keep the variable declaration short and sweet. So I'd recommend rewriting the error handling function to something like this:

```
    private func errorUpdate(for exception: KotlinException) {
        let errorMessage = (exception as? Error)?.localizedDescription ?? "KotlinException to Error casting failed"
        log.e(withMessage: {"Displaying error: \(errorMessage)"})
        let alertController = UIAlertController(title: "error", message: errorMessage, preferredStyle: .actionSheet)
        alertController.addAction(UIAlertAction(title: "Dismiss", style: .default))
        present(alertController, animated: true, completion: nil)
    }
```

LMK if you've got any questions! Or for that matter if you disagree

Reply:
Sounds good, thanks! I'll reformat those. Would be great if we had a lint on the iOS app.

I'm going to refactor `DataState.Error` to take a `String` instead of a Kotlin `Exception`, which will obviate the downcasting problem.